dubbo-registry/dubbo-registry-zookeeper/src/main/java/com/alibaba/dubbo/registry/zookeeper/ZookeeperRegistry.java

```diff
@@ -256,6 +259,23 @@ private String toUrlPath(URL url) {
         return toCategoryPath(url) + Constants.PATH_SEPARATOR + URL.encode(url.toFullString());
     }
 
+    protected List<String> toUnsubscribedPath(URL url){
+        List<String> categories = new ArrayList<String>();
+        if(Constants.ANY_VALUE.equals(url.getServiceInterface())) {
+            String group = url.getParameter(Constants.GROUP_KEY, DEFAULT_ROOT);

```

Comment:
Get the group of registry is redundant when there is a `root` property. 
I think we can use `root` directly by changing `AbstractRegistryFactory` to create a Registry instance for every group:
```java
 public Registry getRegistry(URL url) {
        url = url.setPath(RegistryService.class.getName())
                .addParameter(Constants.INTERFACE_KEY, RegistryService.class.getName())
                .removeParameters(Constants.EXPORT_KEY, Constants.REFER_KEY);
       // Create new Registry instance for different groups.
        String group = url.getParameter(Constants.GROUP_KEY, DEFAULT_ROOT);
        String key = group + Constants.PATH_SEPARATOR + url.toServiceString();
        // Lock the registry access process to ensure a single instance of the registry
        LOCK.lock();
        try {
            Registry registry = REGISTRIES.get(key);
            if (registry != null) {
                return registry;
            }
            registry = createRegistry(url);
            if (registry == null) {
                throw new IllegalStateException("Can not create registry " + url);
            }
            REGISTRIES.put(key, registry);
            return registry;
        } finally {
            // Release the lock
            LOCK.unlock();
        }
    }
```

Reply:
for question 1: Get the group of registry is redundant when there is a root property.
I don't think so, we can't guarantee any interface equals "*" and url's GROUP_KEY is root, in the future maybe changed. 
for question 2: I think we can use root directly by changing AbstractRegistryFactory to create a Registry instance for every group
Sorry, I can  not understand why do that.

Reply:
>for question 1: Get the group of registry is redundant when there is a root property.

>I don't think so, we can't guarantee any interface equals "*" and url's GROUP_KEY is root, in >the future maybe changed.

In which scenario will users change a `Registry Group` dynamically?

>for question 2: I think we can use root directly by changing AbstractRegistryFactory to create >a Registry instance for every group

>Sorry, I can not understand why do that.

I mean users may configure two different Registry Groups in one application, ZookeeperRegistry  can not work properly for that use case: only one Registry instance will be created and one Group take effect.

AbstractRegistryFactory
```java
public Registry getRegistry(URL url) {
        url = url.setPath(RegistryService.class.getName())
                .addParameter(Constants.INTERFACE_KEY, RegistryService.class.getName())
                .removeParameters(Constants.EXPORT_KEY, Constants.REFER_KEY);
        String key = url.toServiceString();
        // Lock the registry access process to ensure a single instance of the registry
        LOCK.lock();
        try {
            Registry registry = REGISTRIES.get(key);
            if (registry != null) {
                return registry;
            }
            registry = createRegistry(url);
            if (registry == null) {
                throw new IllegalStateException("Can not create registry " + url);
            }
            REGISTRIES.put(key, registry);
            return registry;
        } finally {
            // Release the lock
            LOCK.unlock();
        }
    }
```

ZookeeperRegistry
```java
public ZookeeperRegistry(URL url, ZookeeperTransporter zookeeperTransporter) {
        super(url);
        if (url.isAnyHost()) {
            throw new IllegalStateException("registry address == null");
        }
        String group = url.getParameter(Constants.GROUP_KEY, DEFAULT_ROOT);
        if (!group.startsWith(Constants.PATH_SEPARATOR)) {
            group = Constants.PATH_SEPARATOR + group;
        }
        this.root = group;
        zkClient = zookeeperTransporter.connect(url);
        zkClient.addStateListener(new StateListener() {
            public void stateChanged(int state) {
                if (state == RECONNECTED) {
                    try {
                        recover();
                    } catch (Exception e) {
                        logger.error(e.getMessage(), e);
                    }
                }
            }
        });
    }
```